compiler/rustc_hir_typeck/src/fn_ctxt/adjust_fulfillment_errors.rs

```diff
@@ -67,6 +74,33 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             })
         };
 
+        // Account for enum variant constructors, where the type param corresponds to the enum
+        // itself.
+        let enum_def_id = if let hir::def::DefKind::Ctor(hir::def::CtorOf::Variant, _) =
+            self.tcx.def_kind(def_id)
+        {
+            // `def_id` corresponds to a constructor, and its parent is the variant, and we want
+            // the enum.
+            Some(self.tcx.parent(self.tcx.parent(def_id)))
+        } else {
+            None
+        };
+        let variant_param_to_point_at = find_param_matching(&|param_term| {
```

Comment:
Please remove `variant_param_to_point_at`, too. The changes that it has on UI tests like `tests/ui/function-pointer/unsized-ret.stderr` can be achieved without having to add this extra argument -- it's essentially always covered by either `param_to_point_at` or `fallback_param_to_point_at`, depending on what the HIR path looks like.

The correct approach to this is to change the order of `hir::ExprKind::Path(qpath)` to prefer pointing at generics over pointing at args. This can be done by reordering the calls to `point_at_path_if_possible` and `blame_specific_arg_if_possible`, though this is subtle, so I would appreciate if you put some consideration into the tradeoffs here, since this has the side-effect of possibly squashing other suggestions that rely on `FunctionArgumentObligation` that come from `blame_specific_arg_if_possible`.

Reply:
Or, if you don't want to bother with this, I can do it.

Reply:
We still need to differentiate `impl Trait` params, which we don't want to point to for `point_at_generic_if_possible` but we do for `blame_specific_arg_if_possible`.

Reply:
Can you point me to an example where this matters?

Reply:
I reworked this whole thing. I finally realized that the reason I was getting wonky results is because we always do a fuzzy search looking for the right argument, but by adding an explicit check for projections on type parameters we can always have a correct understanding that the problem came from a type parameter, which we can point to when present, *then* look for fn arguments, and finally do the fuzzy search. I'm quite satisfied with the results of that order of operations.